File: Sources/ComicCore/GlyphOrientation.swift
import Foundation
import CoreGraphics
import Vision
import CoreText

public struct GlyphOrientationEvidence: Codable, Sendable {
    public var scores: [String: Double] = [:]
    public var characters: [String: Int] = [:]
    public var rotation: Int?
    public var regions = 0
}

/// Reflow CJK character grids inside enclosed white regions into horizontal OCR lines.
/// The glyph pixels retain their candidate orientation; vertical layout no longer masks it.
public enum GlyphOrientation {
    private struct Mask { var width:Int; var height:Int; var pixels:[UInt8]; var originX=0; var originY=0 }
    private struct Component {
        var ink:Bool;var x:Int;var y:Int;var right:Int;var bottom:Int;var points:[Int]
        var area:Int {(right-x)*(bottom-y)}
    }
    public static func reviewSheets(_ image:CGImage) throws -> [Int:CGImage] {
        let masks=try regions(image)
        return Dictionary(uniqueKeysWithValues:[0,90,180,270].compactMap { angle in
            sheet(masks.compactMap{line(rotate($0,angle:angle))}).map{(angle,$0)}
        })
    }
    public static func analyze(_ image:CGImage) throws -> GlyphOrientationEvidence {
        let masks=try regions(image)
        var evidence=GlyphOrientationEvidence();evidence.regions=masks.count
        guard !masks.isEmpty else {return evidence}
        // OCR may read sideways CJK glyphs while reporting a horizontal text box.
        // Validate the physical strokes against rendered candidates, not OCR confidence.
        struct Vote {var region:Int;var bounds:CGRect;var rotation:Int;var score:Double;var margin:Double}
        var votes:[Vote]=[]
        for angle in [0,90,180,270] {
            try Task.checkCancellation()
            let lines=masks.enumerated().compactMap { index,mask in line(rotate(mask,angle:angle)).map{(index,$0)} }
            guard let sheet=sheet(lines.map{$0.1}) else {continue}
            let request=VNRecognizeTextRequest();request.revision=VNRecognizeTextRequestRevision3
            request.recognitionLevel = .accurate;request.usesLanguageCorrection=false;request.minimumTextHeight=0
            request.recognitionLanguages=["zh-Hant","en-US"]
            try VNImageRequestHandler(cgImage:sheet).perform([request])
            for observation in request.results ?? [] {
                guard let candidate=observation.topCandidates(1).first,candidate.confidence>=0.3 else {continue}
                let text=candidate.string
                for index in text.indices {
                    let character=String(text[index])
                    guard character.unicodeScalars.contains(where:{(0x3400...0x9fff).contains($0.value)}),
                          let rectangle=try? candidate.boundingBox(for:index..<text.index(after:index)) else {continue}
                    let center=rectangle.boundingBox
                    let column=Int((Double(center.midX)*Double(sheet.width)-12.0)/48.0)
                    let row=Int(((1.0-Double(center.midY))*Double(sheet.height)-12.0)/64.0)
                    guard lines.indices.contains(row),lines[row].1.indices.contains(column) else {continue}
                    let actual=normalized(lines[row].1[column])
                    let templates=templates(character)
                    var matches:[(Int,Double)]=[]
                    for turn in 0..<4 {
                        let score=templates.map{similarity(actual,rotateSquare($0,turn:turn))}.max() ?? 0
                        matches.append((turn*90,score))
                    }
                    matches.sort{$0.1>$1.1}
                    let best=matches[0],margin=best.1-matches[1].1
                    guard best.1>=0.55,margin>=0.035 else {continue}
                    let region=lines[row].0,tile=lines[row].1[column],original=masks[region]
                    let bounds: CGRect
                    switch angle {
                    case 90: bounds=CGRect(x:tile.originY,y:original.height-tile.originX-tile.width,width:tile.height,height:tile.width)
                    case 180: bounds=CGRect(x:original.width-tile.originX-tile.width,y:original.height-tile.originY-tile.height,width:tile.width,height:tile.height)
                    case 270: bounds=CGRect(x:original.width-tile.originY-tile.height,y:tile.originX,width:tile.height,height:tile.width)
                    default: bounds=CGRect(x:tile.originX,y:tile.originY,width:tile.width,height:tile.height)
                    }
                    let vote=Vote(region:region,bounds:bounds,rotation:(angle-best.0+360)%360,score:best.1,margin:margin)
                    if let existing=votes.firstIndex(where:{ value in
                        let overlap=value.bounds.intersection(bounds)
                        return value.region==region && !overlap.isNull && overlap.width*overlap.height>min(value.bounds.width*value.bounds.height,bounds.width*bounds.height)*0.6
                    }) {
                        if margin*best.1>votes[existing].margin*votes[existing].score {votes[existing]=vote}
                    } else {votes.append(vote)}
                }
            }
        }
        for vote in votes {
            evidence.scores[String(vote.rotation),default:0]+=vote.score
            evidence.characters[String(vote.rotation),default:0]+=1
        }
        let ranked=evidence.scores.sorted{$0.value>$1.value}
        if let best=ranked.first {
            let runner=ranked.dropFirst().first?.value ?? 0
            if evidence.characters[best.key,default:0]>=3,best.value>=2.4,
               best.value-runner>=1.8,best.value>=max(0.1,runner)*1.6 {
                evidence.rotation=Int(best.key)
            }
        }
        return evidence
    }
    private final class Templates: NSObject {let values:[[UInt8]];init(_ values:[[UInt8]]) {self.values=values}}
    private static let templateCache:NSCache<NSString,Templates> = {
        let cache=NSCache<NSString,Templates>();cache.countLimit=1024;return cache
    }()
    private static func templates(_ character:String)->[[UInt8]] {
        if let cached=templateCache.object(forKey:character as NSString) {return cached.values}
        let values=["PingFangTC-Semibold","STHeitiTC-Light","STKaiti","SongtiTC-Bold"].compactMap { name -> [UInt8]? in
            var pixels=[UInt8](repeating:255,count:80*80)
            let result=pixels.withUnsafeMutableBytes { bytes -> Bool in
                guard let context=CGContext(data:bytes.baseAddress,width:80,height:80,bitsPerComponent:8,bytesPerRow:80,space:CGColorSpaceCreateDeviceGray(),bitmapInfo:0) else {return false}
                context.setFillColor(CGColor(gray:1,alpha:1));context.fill(CGRect(x:0,y:0,width:80,height:80))
                let font=CTFontCreateWithName(name as CFString,60,nil)
                let string=NSAttributedString(string:character,attributes:[.init(kCTFontAttributeName as String):font,.init(kCTForegroundColorAttributeName as String):CGColor(gray:0,alpha:1)])
                context.textPosition=CGPoint(x:8,y:15);CTLineDraw(CTLineCreateWithAttributedString(string),context);return true
            }
            guard result else {return nil}
            return normalized(Mask(width:80,height:80,pixels:pixels.map{$0<205 ? 0 : 255}))
        }
        templateCache.setObject(Templates(values),forKey:character as NSString);return values
    }
    private static func normalized(_ mask:Mask)->[UInt8] {
        var minX=mask.width,minY=mask.height,maxX=0,maxY=0
        for y in 0..<mask.height {for x in 0..<mask.width where mask.pixels[y*mask.width+x]<128 {
            minX=min(minX,x);minY=min(minY,y);maxX=max(maxX,x);maxY=max(maxY,y)
        }}
        guard minX<=maxX,minY<=maxY else {return [UInt8](repeating:0,count:1024)}
        var result=[UInt8](repeating:0,count:1024)
        for y in 0..<32 {for x in 0..<32 {
            let sx=minX+min(maxX-minX,Int((Double(x)+0.5)*Double(maxX-minX+1)/32))
            let sy=minY+min(maxY-minY,Int((Double(y)+0.5)*Double(maxY-minY+1)/32))
            result[y*32+x]=mask.pixels[sy*mask.width+sx]<128 ? 1 : 0
        }}
        return result
    }
    private static func rotateSquare(_ pixels:[UInt8],turn:Int)->[UInt8] {
        if turn==0 {return pixels}
        var result=[UInt8](repeating:0,count:1024)
        for y in 0..<32 {for x in 0..<32 {
            let nx=turn==1 ? 31-y : (turn==2 ? 31-x : y)
            let ny=turn==1 ? x : (turn==2 ? 31-y : 31-x)
            result[ny*32+nx]=pixels[y*32+x]
        }}
        return result
    }
    private static func similarity(_ a:[UInt8],_ b:[UInt8])->Double {
        var intersection=0,total=0
        for i in a.indices {intersection+=Int(a[i]&b[i]);total+=Int(a[i])+Int(b[i])}
        return Double(2*intersection)/Double(max(1,total))
    }
    private static func regions(_ image:CGImage) throws -> [Mask] {
        let factor=min(1,1000.0/Double(max(image.width,image.height)))
        let w=Int(Double(image.width)*factor),h=Int(Double(image.height)*factor)
        var pixels=[UInt8](repeating:255,count:w*h)
        let ok=pixels.withUnsafeMutableBytes { bytes -> Bool in
            guard let context=CGContext(data:bytes.baseAddress,width:w,height:h,bitsPerComponent:8,bytesPerRow:w,space:CGColorSpaceCreateDeviceGray(),bitmapInfo:0) else {return false}
            context.setFillColor(CGColor(gray:1,alpha:1));context.fill(CGRect(x:0,y:0,width:w,height:h))
            context.draw(image,in:CGRect(x:0,y:0,width:w,height:h));return true
        }
        guard ok else {return []}
        var labels=[Int](repeating:-1,count:w*h),components:[Component]=[]
        func neighbors(_ p:Int)->[Int] {
            let x=p%w,y=p/w
            return [x>0 ? p-1 : -1,x+1<w ? p+1 : -1,y>0 ? p-w : -1,y+1<h ? p+w : -1].filter{$0>=0}
        }
        for p in pixels.indices where labels[p]<0 {
            if p%w==0 {try Task.checkCancellation()}
            let ink=pixels[p]<205,id=components.count
            var component=Component(ink:ink,x:p%w,y:p/w,right:p%w+1,bottom:p/w+1,points:[p]);labels[p]=id
            var head=0
            while head<component.points.count {
                let q=component.points[head];head+=1
                let x=q%w,y=q/w
                component.x=min(component.x,x);component.y=min(component.y,y);component.right=max(component.right,x+1);component.bottom=max(component.bottom,y+1)
                for next in neighbors(q) where labels[next]<0 && (pixels[next]<205)==ink {labels[next]=id;component.points.append(next)}
            }
            components.append(component)
        }
        var groups:[Int:[Int]]=[:]
        for (id,c) in components.enumerated() where c.ink && c.points.count>=2 && c.right-c.x<w*15/100 && c.bottom-c.y<h*15/100 {
            var adjacent:[Int:Int]=[:]
            for p in c.points {for n in neighbors(p) where pixels[n]>=205 {adjacent[labels[n],default:0]+=1}}
            guard let parent=adjacent.max(by:{$0.value<$1.value})?.key else {continue}
            let area=components[parent]
            guard area.points.count>=300,Double(area.points.count)/Double(area.area)>=0.5,area.area<w*h/4,
                  c.x>area.x,c.y>area.y,c.right<area.right,c.bottom<area.bottom else {continue}
            groups[parent,default:[]].append(id)
        }
        var masks:[Mask]=[]
        for key in groups.keys.sorted() {
            let ids=groups[key]!
            guard ids.count>=4 else {continue}
            let parts=ids.map{components[$0]}
            let x=parts.map(\.x).min()!,y=parts.map(\.y).min()!,right=parts.map(\.right).max()!,bottom=parts.map(\.bottom).max()!
            let width=right-x,height=bottom-y
            guard width>=10,height>=10 else {continue}
            var mask=Mask(width:width,height:height,pixels:[UInt8](repeating:255,count:width*height))
            for part in parts {for p in part.points {mask.pixels[(p/w-y)*width+p%w-x]=0}}
            if line(mask) != nil {masks.append(mask)}
        }
        // Bound OCR cost and give larger text grids priority over tiny decoration.
        return Array(masks.sorted{$0.width*$0.height>$1.width*$1.height}.prefix(12))
    }
    private static func rotate(_ mask:Mask,angle:Int)->Mask {
        if angle==0 {return mask}
        let w=angle==180 ? mask.width : mask.height,h=angle==180 ? mask.height : mask.width
        var result=Mask(width:w,height:h,pixels:[UInt8](repeating:255,count:w*h))
        for y in 0..<mask.height {for x in 0..<mask.width {
            let nx=angle==90 ? mask.height-1-y : (angle==180 ? mask.width-1-x : y)
            let ny=angle==90 ? x : (angle==180 ? mask.height-1-y : mask.width-1-x)
            result.pixels[ny*w+nx]=mask.pixels[y*mask.width+x]
        }}
        return result
    }
    private static func bands(_ projection:[Bool])->[Range<Int>] {
        var result:[Range<Int>]=[];var start:Int?
        for i in projection.indices {
            if projection[i],start==nil {start=i}
            if !projection[i],let first=start {result.append(first..<i);start=nil}
        }
        if let first=start {result.append(first..<projection.count)}
        return result
    }
    private static func line(_ mask:Mask)->[Mask]? {
        var xs=[Bool](repeating:false,count:mask.width),ys=[Bool](repeating:false,count:mask.height)
        for y in 0..<mask.height {for x in 0..<mask.width where mask.pixels[y*mask.width+x]==0 {xs[x]=true;ys[y]=true}}
        let xb=bands(xs),yb=bands(ys)
        let sizes=(xb+yb).map(\.count).filter{$0>=5}.sorted()
        guard !sizes.isEmpty else {return nil}
        let pitch=Double(sizes[sizes.count/2])
        func merged(_ bands:[Range<Int>])->[Range<Int>] {
            var result:[Range<Int>]=[]
            for band in bands {
                if let previous=result.last,Double(previous.count)<pitch*0.65,Double(band.upperBound-previous.lowerBound)<=pitch*1.4 {
                    result[result.count-1]=previous.lowerBound..<band.upperBound
                } else {result.append(band)}
            }
            return result
        }
        var glyphs:[Mask]=[]
        for x in merged(xb).reversed() {for y in merged(yb) {
            guard Double(x.count)/Double(y.count)>=0.6,Double(x.count)/Double(y.count)<=1.65,
                  Double(min(x.count,y.count))>=pitch*0.6 else {continue}
            var pixels:[UInt8]=[]
            for row in y {pixels.append(contentsOf:mask.pixels[(row*mask.width+x.lowerBound)..<(row*mask.width+x.upperBound)])}
            let coverage=Double(pixels.filter{$0==0}.count)/Double(pixels.count)
            guard coverage>=0.08,coverage<=0.80 else {continue}
            glyphs.append(Mask(width:x.count,height:y.count,pixels:pixels,originX:x.lowerBound,originY:y.lowerBound))
        }}
        return glyphs.count>=3 && glyphs.count<=24 ? glyphs : nil
    }
    private static func sheet(_ lines:[[Mask]])->CGImage? {
        guard !lines.isEmpty else {return nil}
        let cell=48,w=24+cell*(lines.map(\.count).max() ?? 1),h=24+lines.count*64
        guard let context=CGContext(data:nil,width:w,height:h,bitsPerComponent:8,bytesPerRow:0,space:CGColorSpaceCreateDeviceGray(),bitmapInfo:0) else {return nil}
        context.setFillColor(CGColor(gray:1,alpha:1));context.fill(CGRect(x:0,y:0,width:w,height:h))
        for (row,glyphs) in lines.enumerated() {for (column,glyph) in glyphs.enumerated() {
            guard let provider=CGDataProvider(data:Data(glyph.pixels) as CFData),let image=CGImage(width:glyph.width,height:glyph.height,bitsPerComponent:8,bitsPerPixel:8,bytesPerRow:glyph.width,space:CGColorSpaceCreateDeviceGray(),bitmapInfo:CGBitmapInfo(rawValue:0),provider:provider,decode:nil,shouldInterpolate:true,intent:.defaultIntent) else {continue}
            context.draw(image,in:CGRect(x:12+column*cell,y:h-12-(row+1)*64,width:40,height:40))
        }}
        return context.makeImage()
    }
}

File: android/app/src/main/java/com/mandu/reader/core/GlyphOrientation.kt
package com.mandu.reader.core

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import android.util.LruCache
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.util.Locale
import kotlin.coroutines.coroutineContext
import kotlin.math.floor

/**
 * Port of ComicCore's local bubble/glyph orientation evidence. Reflow changes the
 * layout, never the glyph strokes. Thus upright vertical CJK is upright evidence.
 * Returned angles are clockwise display corrections in Android image coordinates.
 */
object GlyphOrientation {
    data class Evidence(
        val rotation: Int? = null,
        val scores: Map<Int, Double> = emptyMap(),
        val characters: Map<Int, Int> = emptyMap(),
        val regions: Int = 0
    )

    private data class Mask(val width: Int, val height: Int, val pixels: ByteArray, val originX: Int = 0, val originY: Int = 0)
    private data class Component(val ink: Boolean, val bounds: Rect, val points: IntArray)
    private data class Vote(val region: Int, val bounds: Rect, val rotation: Int, val score: Double, val margin: Double)
    private data class GlyphLine(val region: Int, val glyphs: List<Mask>)
    private val templateCache = LruCache<String, List<ByteArray>>(1024)
    private val angles = listOf(0, 90, 180, 270)

    suspend fun analyze(bitmap: Bitmap): Evidence = withContext(Dispatchers.Default) {
        val masks = regions(bitmap)
        if (masks.isEmpty()) return@withContext Evidence()
        val recognizer = TextRecognition.getClient(JapaneseTextRecognizerOptions.Builder().build())
        val votes = mutableListOf<Vote>()
        try {
            for (angle in angles) {
                coroutineContext.ensureActive()
                val lines = masks.mapIndexedNotNull { index, mask -> line(rotate(mask, angle))?.let { GlyphLine(index, it) } }
                if (lines.isEmpty()) continue
                val sheet = sheet(lines)
                try {
                    val text = recognizer.process(InputImage.fromBitmap(sheet, 0)).await()
                    for (block in text.textBlocks) for (textLine in block.lines) for (element in textLine.elements) {
                        for (symbol in element.symbols) {
                            val character = symbol.text
                            if (character.length != 1 || character[0].code !in 0x3400..0x9fff || symbol.confidence < .3f) continue
                            val box = symbol.boundingBox ?: continue
                            val column = floor((box.exactCenterX() - 12) / 48.0).toInt()
                            val row = floor((box.exactCenterY() - 12) / 64.0).toInt()
                            val glyphLine = lines.getOrNull(row) ?: continue
                            val tile = glyphLine.glyphs.getOrNull(column) ?: continue
                            val actual = normalized(tile)
                            val templates = templates(character)
                            val matches = (0..3).map { turn ->
                                turn * 90 to (templates.maxOfOrNull { similarity(actual, rotateSquare(it, turn)) } ?: 0.0)
                            }.sortedByDescending { it.second }
                            val best = matches[0]
                            val margin = best.second - matches[1].second
                            if (best.second < .55 || margin < .035) continue
                            val original = masks[glyphLine.region]
                            val bounds = originalBounds(tile, original, angle)
                            val vote = Vote(glyphLine.region, bounds, (angle - best.first + 360) % 360, best.second, margin)
                            // Four candidate sheets can recognize one physical glyph four times.
                            // Keep just its strongest stroke match, even when the OCR labels differ.
                            val existing = votes.indexOfFirst { it.region == vote.region && overlaps(it.bounds, bounds) }
                            if (existing < 0) votes += vote
                            else if (vote.margin * vote.score > votes[existing].margin * votes[existing].score) votes[existing] = vote
                        }
                    }
                } finally {
                    sheet.recycle()
                }
            }
        } finally {
            recognizer.close()
        }
        val scores = mutableMapOf<Int, Double>()
        val characters = mutableMapOf<Int, Int>()
        for (vote in votes) {
            scores[vote.rotation] = (scores[vote.rotation] ?: 0.0) + vote.score
            characters[vote.rotation] = (characters[vote.rotation] ?: 0) + 1
        }
        val ranked = scores.entries.sortedByDescending { it.value }
        val best = ranked.firstOrNull()
        val runner = ranked.getOrNull(1)?.value ?: 0.0
        val rotation = best?.takeIf {
            (characters[it.key] ?: 0) >= 3 && it.value >= 2.4 && it.value - runner >= 1.8 && it.value >= maxOf(.1, runner) * 1.6
        }?.key
        Evidence(rotation, scores.toMap(), characters.toMap(), masks.size)
    }

    private fun originalBounds(tile: Mask, original: Mask, angle: Int): Rect {
        val x: Int; val y: Int; val width: Int; val height: Int
        when (angle) {
            90 -> { x = tile.originY; y = original.height - tile.originX - tile.width; width = tile.height; height = tile.width }
            180 -> { x = original.width - tile.originX - tile.width; y = original.height - tile.originY - tile.height; width = tile.width; height = tile.height }
            270 -> { x = original.width - tile.originY - tile.height; y = tile.originX; width = tile.height; height = tile.width }
            else -> { x = tile.originX; y = tile.originY; width = tile.width; height = tile.height }
        }
        return Rect(x, y, x + width, y + height)
    }

    private fun overlaps(a: Rect, b: Rect): Boolean {
        val width = minOf(a.right, b.right) - maxOf(a.left, b.left)
        val height = minOf(a.bottom, b.bottom) - maxOf(a.top, b.top)
        return width > 0 && height > 0 && width * height > minOf(a.width() * a.height(), b.width() * b.height()) * .6
    }

    private fun templates(character: String): List<ByteArray> {
        templateCache.get(character)?.let { return it }
        val variants = mutableListOf<ByteArray>()
        for (locale in listOf(Locale.JAPAN, Locale.TRADITIONAL_CHINESE)) {
            for ((family, style) in listOf("sans-serif" to Typeface.NORMAL, "sans-serif" to Typeface.BOLD, "serif" to Typeface.NORMAL, "serif" to Typeface.BOLD)) {
                val bitmap = Bitmap.createBitmap(80, 80, Bitmap.Config.ARGB_8888)
                try {
                    val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                        color = Color.BLACK; textSize = 60f; typeface = Typeface.create(family, style); textLocale = locale
                    }
                    if (!paint.hasGlyph(character)) continue
                    val bounds = Rect()
                    paint.getTextBounds(character, 0, character.length, bounds)
                    val canvas = Canvas(bitmap)
                    canvas.drawColor(Color.WHITE)
                    canvas.drawText(character, (80 - bounds.width()) / 2f - bounds.left, (80 - bounds.height()) / 2f - bounds.top, paint)
                    variants += normalized(binary(bitmap))
                } finally { bitmap.recycle() }
            }
        }
        templateCache.put(character, variants)
        return variants
    }

    private fun normalized(mask: Mask): ByteArray {
        var minX = mask.width; var minY = mask.height; var maxX = -1; var maxY = -1
        for (y in 0 until mask.height) for (x in 0 until mask.width) if (mask.pixels[y * mask.width + x] == 1.toByte()) {
            minX = minOf(minX, x); minY = minOf(minY, y); maxX = maxOf(maxX, x); maxY = maxOf(maxY, y)
        }
        val result = ByteArray(1024)
        if (maxX < minX || maxY < minY) return result
        for (y in 0 until 32) for (x in 0 until 32) {
            val sx = minX + minOf(maxX - minX, ((x + .5) * (maxX - minX + 1) / 32).toInt())
            val sy = minY + minOf(maxY - minY, ((y + .5) * (maxY - minY + 1) / 32).toInt())
            result[y * 32 + x] = mask.pixels[sy * mask.width + sx]
        }
        return result
    }

    private fun rotateSquare(pixels: ByteArray, turn: Int): ByteArray {
        if (turn == 0) return pixels
        val result = ByteArray(1024)
        for (y in 0 until 32) for (x in 0 until 32) {
            val nx = when (turn) { 1 -> 31 - y; 2 -> 31 - x; else -> y }
            val ny = when (turn) { 1 -> x; 2 -> 31 - y; else -> 31 - x }
            result[ny * 32 + nx] = pixels[y * 32 + x]
        }
        return result
    }

    private fun similarity(a: ByteArray, b: ByteArray): Double {
        var intersection = 0; var total = 0
        for (i in a.indices) { intersection += a[i].toInt() and b[i].toInt(); total += a[i] + b[i] }
        return 2.0 * intersection / maxOf(1, total)
    }

    private fun binary(bitmap: Bitmap): Mask {
        val colors = IntArray(bitmap.width * bitmap.height)
        bitmap.getPixels(colors, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
        val pixels = ByteArray(colors.size) { i ->
            val c = colors[i]
            val luminance = (Color.red(c) * 299 + Color.green(c) * 587 + Color.blue(c) * 114) / 1000
            val whiteComposite = (luminance * Color.alpha(c) + 255 * (255 - Color.alpha(c))) / 255
            if (whiteComposite < 205) 1 else 0
        }
        return Mask(bitmap.width, bitmap.height, pixels)
    }

    private suspend fun regions(image: Bitmap): List<Mask> {
        val factor = minOf(1.0, 1000.0 / maxOf(image.width, image.height))
        val scaled = Bitmap.createScaledBitmap(image, maxOf(1, (image.width * factor).toInt()), maxOf(1, (image.height * factor).toInt()), true)
        val mask = try { binary(scaled) } finally { if (scaled !== image) scaled.recycle() }
        val w = mask.width; val h = mask.height; val pixels = mask.pixels
        val labels = IntArray(pixels.size) { -1 }
        val queue = IntArray(pixels.size)
        val components = mutableListOf<Component>()
        for (p in pixels.indices) {
            if (p % w == 0) coroutineContext.ensureActive()
            if (labels[p] >= 0) continue
            val ink = pixels[p] == 1.toByte()
            val id = components.size
            var head = 0; var tail = 1
            queue[0] = p; labels[p] = id
            var x = p % w; var y = p / w; var right = x + 1; var bottom = y + 1
            while (head < tail) {
                if (head % 16384 == 0) coroutineContext.ensureActive()
                val q = queue[head++]
                x = minOf(x, q % w); y = minOf(y, q / w); right = maxOf(right, q % w + 1); bottom = maxOf(bottom, q / w + 1)
                neighbors(q, w, h) { next ->
                    if (labels[next] < 0 && pixels[next] == pixels[p]) { labels[next] = id; queue[tail++] = next }
                }
            }
            components += Component(ink, Rect(x, y, right, bottom), queue.copyOf(tail))
        }
        val groups = mutableMapOf<Int, MutableList<Int>>()
        for ((id, c) in components.withIndex()) {
            coroutineContext.ensureActive()
            if (!c.ink || c.points.size < 2 || c.bounds.width() >= w * 15 / 100 || c.bounds.height() >= h * 15 / 100) continue
            val adjacent = mutableMapOf<Int, Int>()
            for (p in c.points) neighbors(p, w, h) { n ->
                if (pixels[n] == 0.toByte()) adjacent[labels[n]] = (adjacent[labels[n]] ?: 0) + 1
            }
            val parent = adjacent.maxByOrNull { it.value }?.key ?: continue
            val area = components[parent]
            val bounds = area.bounds
            val boxArea = bounds.width() * bounds.height()
            if (area.points.size < 300 || area.points.size.toDouble() / boxArea < .5 || boxArea >= w * h / 4 ||
                c.bounds.left <= bounds.left || c.bounds.top <= bounds.top || c.bounds.right >= bounds.right || c.bounds.bottom >= bounds.bottom) continue
            groups.getOrPut(parent) { mutableListOf() } += id
        }
        val masks = mutableListOf<Mask>()
        for (key in groups.keys.sorted()) {
            val ids = groups.getValue(key)
            if (ids.size < 4) continue
            val parts = ids.map { components[it] }
            val x = parts.minOf { it.bounds.left }; val y = parts.minOf { it.bounds.top }
            val width = parts.maxOf { it.bounds.right } - x; val height = parts.maxOf { it.bounds.bottom } - y
            if (width < 10 || height < 10) continue
            val tile = Mask(width, height, ByteArray(width * height))
            for (part in parts) for (p in part.points) tile.pixels[(p / w - y) * width + p % w - x] = 1
            if (line(tile) != null) masks += tile
        }
        return masks.sortedByDescending { it.width * it.height }.take(12)
    }

    private inline fun neighbors(p: Int, w: Int, h: Int, action: (Int) -> Unit) {
        if (p % w > 0) action(p - 1)
        if (p % w + 1 < w) action(p + 1)
        if (p >= w) action(p - w)
        if (p < w * (h - 1)) action(p + w)
    }

    private fun rotate(mask: Mask, angle: Int): Mask {
        if (angle == 0) return mask
        val w = if (angle == 180) mask.width else mask.height
        val h = if (angle == 180) mask.height else mask.width
        val result = Mask(w, h, ByteArray(w * h))
        for (y in 0 until mask.height) for (x in 0 until mask.width) {
            val nx = when (angle) { 90 -> mask.height - 1 - y; 180 -> mask.width - 1 - x; else -> y }
            val ny = when (angle) { 90 -> x; 180 -> mask.height - 1 - y; else -> mask.width - 1 - x }
            result.pixels[ny * w + nx] = mask.pixels[y * mask.width + x]
        }
        return result
    }

    private fun bands(projection: BooleanArray): List<IntRange> {
        val result = mutableListOf<IntRange>()
        var start = -1
        for (i in projection.indices) {
            if (projection[i] && start < 0) start = i
            if (!projection[i] && start >= 0) { result += start until i; start = -1 }
        }
        if (start >= 0) result += start until projection.size
        return result
    }

    private fun line(mask: Mask): List<Mask>? {
        val xs = BooleanArray(mask.width); val ys = BooleanArray(mask.height)
        for (y in 0 until mask.height) for (x in 0 until mask.width) if (mask.pixels[y * mask.width + x] == 1.toByte()) { xs[x] = true; ys[y] = true }
        val xb = bands(xs); val yb = bands(ys)
        val sizes = (xb + yb).map { it.last - it.first + 1 }.filter { it >= 5 }.sorted()
        if (sizes.isEmpty()) return null
        val pitch = sizes[sizes.size / 2].toDouble()
        fun merged(bands: List<IntRange>): List<IntRange> {
            val result = mutableListOf<IntRange>()
            for (band in bands) {
                val previous = result.lastOrNull()
                if (previous != null && previous.last - previous.first + 1 < pitch * .65 && band.last - previous.first + 1 <= pitch * 1.4) {
                    result[result.lastIndex] = previous.first..band.last
                } else result += band
            }
            return result
        }
        val glyphs = mutableListOf<Mask>()
        for (x in merged(xb).reversed()) for (y in merged(yb)) {
            val width = x.last - x.first + 1; val height = y.last - y.first + 1
            if (width.toDouble() / height !in .6..1.65 || minOf(width, height) < pitch * .6) continue
            val pixels = ByteArray(width * height)
            for (row in y) System.arraycopy(mask.pixels, row * mask.width + x.first, pixels, (row - y.first) * width, width)
            val coverage = pixels.count { it == 1.toByte() }.toDouble() / pixels.size
            if (coverage !in .08.. .80) continue
            glyphs += Mask(width, height, pixels, x.first, y.first)
        }
        return glyphs.takeIf { it.size in 3..24 }
    }

    private fun sheet(lines: List<GlyphLine>): Bitmap {
        val result = Bitmap.createBitmap(24 + 48 * lines.maxOf { it.glyphs.size }, 24 + lines.size * 64, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(result)
        canvas.drawColor(Color.WHITE)
        val paint = Paint(Paint.FILTER_BITMAP_FLAG)
        for ((row, line) in lines.withIndex()) for ((column, glyph) in line.glyphs.withIndex()) {
            val colors = IntArray(glyph.pixels.size) { if (glyph.pixels[it] == 1.toByte()) Color.BLACK else Color.WHITE }
            val tile = Bitmap.createBitmap(colors, glyph.width, glyph.height, Bitmap.Config.ARGB_8888)
            try {
                val left = 12 + column * 48; val top = 12 + row * 64
                canvas.drawBitmap(tile, null, Rect(left, top, left + 40, top + 40), paint)
            } finally { tile.recycle() }
        }
        return result
    }
}
